### ISSUE TITLE
rev746, Texplorer Mips Field?

### ISSUE THREAD

**Issue by ghost** [ME3Explorer/ME3Explorer#276]
I hate to suggest this, b/c A-it's a GUI thing, and B-it's not technically a bug.

However, when we removed the ability to select a specific mipmap to extract and made it just extract the largest (which is entirely logical) we _also_ removed the ability to know with 100% certainty, whether a game texture actually used mipmaps. This is the exact dialog I used to use to confirm whether a modded texture was following the game's mip conventions. I'm currently trying to figure out if I need to generate mips for a normal I need to test, and discovered I have no idea if the game uses mips for it or not. I assume so, but I can't verify this with 100% certainty -- at least not in Texplorer.

So... it seems that we need to add a field to Texplorer, and I'm sure this is much more complex than just  adding a field, since it's something that would affect data collection during treescanning. I don't think it's necessary that we know the quantity of mips, since it's a general rule if mips are present they should go down to 1x1. However, I think it's necessary for us to know whether or not they are present via a simple Yes/No field.

Of course you can simply try to replace a texture and have Texplorer just throw an error at you, but that's a crappy solution, and it assumes Texplorer will be correct -- which it occasionally is not.

**Comment by starkgate** [comment]
If you want to know whether there are MIPs or not, you might as well add the option to know how many there are. It's more understandable than a simple Yes/No, that seems unintuitive to me.

Btw, does anyone know why some MIPs go down to 3 4x4s instead of 4x4 followed by 2x2 and 1x1 ?

**Comment by ghost** [comment]
@CreeperLava -- Presumably, it would be less labor-intensive to implement a simple yes/no field. If not, then K can decide what he wants to do. Like I said, _if_ mips are present, they should go to 1x1, which means you already know how many there are just by looking at the stated size of the texture (largest mip). Putting the number there is more explicit, but if a "y/n" is easier, then the user still actually has the same amount of information... they just need to know how to count, lol.

As far as why some only say they go down to 4x4 or whatnot, I've always viewed that with a bit of skepticism. It seems wrong, though I guess it's something that BW could have done on purpose.

**Comment by alvarome** [comment]
Hi:

Yes, I agree that to extract the largest image is indeed the logical path -or the most usual choice, I bet-, but I´m with CreeperLava here: the more information, the better. Always! ;)

Mips are a mess... not an issue here but, in example, many often you have a texture with, say, 10 mips but only 7 are stored in game... and without detailed info you´ll never know where the cut is. I would tend to think that three smallest are left behind but who knows for sure.

HTH!
Alvaro

**Comment by ghost** [comment]
@alvarome -- I think you're confusing what Creeper is suggesting a little bit (well, or I am!).

He suggested showing the _total number_ of mips. As in "11". This isn't going to tell you _which_ mips are present (e.g., 4x4). If K wants to implement that he certainly can, but it will be even more work than what Creeper and I were suggesting. It would also add a lot of clutter to the Properties panel, that isn't 100% necessary.

Also, to be clear, not every texture is "missing" the last 3 mips. I'd say there is no hard and fast rule. Using Texplorer, I can -- inside of 30s -- find examples of textures that go down to 1x1, ones that go down to 4x4, and ones that have multiple copies of the same mip size. None of these situations are rare, and I'd say that the latter two should be viewed with some skepticism as to whether Texplorer's data is accurate. But ultimately, it doesn't matter. Texplorer is made so that all textures with mips must go down to 1x1, and this is also how photo editors auto-generate mips. Whether the game only goes down to 4x4, or whether it happens to have multiple copies of 4x4 is irrelevant -- we already know replacing down to 1x1 in-game works. 

This is why I said what we really _need_ to know is whether the texture _has_ mips. From there it's just a matter of texture authors ticking "Generate mipmaps" in the image editor of their choice when saving their new texture.

**Comment by KFreon** [comment]
Some background in the game mips vs real mips:
- Mipmaps are usually powers of 2 smaller sizes of the main image. Improves performance when too far away to see details, it can just load a lower detailed version.
- Generally mips go all the way down to 1x1, but they don't need to most of the time.
- The game sometimes doesn't have all the mip levels it should, maybe cos they didn't want to, or maybe Texplorer misses them for other reasons, like they're invalid or something.
- The 4x4 copies may be due to how ImageEngine deals with them, since it can't deal with DXT textures smaller than 4x4.
- The game also seems to have redirects or something for some mipmaps i.e. The info for the image says it has 12 mips, but the top 4 are null for some reason. Note that these nulls are stripped out for that selection window, as will they be for any mip count I implement.

When I do this, it'll be the number of non null mipmaps. I actually thought I had that in there somewhere. I'd look but then I'd need to do the UI fix again :P

**Comment by alvarome** [comment]
Hi:

@giftfish sorry for the confusion here. 

I was refering that in previous revs Texplorer extraction process already gave us all that detailed info. I mean, number of mips and size. Now that is gone.

Perhaps it could be easiest to roll back to the previous texture extraction code than to add a new (or two) fields with that info. 

We would not get the texture mips info displayed as part of the texture properties but it would be just one clic away... at the "extract" button  -?-

HTH!
Alvaro

**Comment by ghost** [comment]
@KFreon -- We need to keep this simple for users. Not everyone is going to have a firm understanding of mipmaps, and we can't force them to read the wiki.

If we start providing numbers that don't match what's auto-generated in PS/GIMP, they're going to be confused. I really think a simple yes/no will suffice. People don't create mipmaps manually. They check a box in the photo editor to tell it whether to generate them. All they need to know is _whether_ to check the box -- which means all they need to know is _whether_ a texture has mips. 

Finally, we'll also generate more confusion if Texplorer starts giving different information than TPF Tools, which gives the normal, full, "null-including" number from your explanation (i.e., 512=10 mips, 1024=11 mips, 2048=12 mips, etc). These numbers have to match across tools, and they have to match the default number auto-generated by the photo editor.

@alvarome -- The old dialog is no longer used b/c it is suboptimal; there's no reason to allow folks to choose. Reintegrating this information into Texplorer requires that it be placed in the Properties, where all other technical info about the texture is. In short, folks shouldn't need to hit extract to know whether a texture has mips.

**Comment by starkgate** [comment]
"Keeping things simple" is not necessarily a good idea. People that use Texplorer are generally modders, normal users will use TPF Tools. Modders should all know what MIP maps are, it's not that hard.
Although I agree with the fact that Texplorer doesn't always display very reliable numbers, we should at least include the same information as TPF Tools : display the number of MIPs following the exponent of 2 rule, no need to actually count them.

That way all tools display the same thing, and I won't have the feeling that something is missing every time I open Texplorer^^. If I have to use an older rev to get information I need, it's not good.

**Comment by ghost** [comment]
@CreeperLava -- I have no problems with displaying the total number, as long as it's _accurate_,  consistent across tools, _and is consistent with image editors' auto-generate mips feature_. 

K has basically already underscored my point that the data Texplorer was presenting information about mips in the old dialog wasn't 100% accurate. This is all the data people have to go on when extracting a texture and modifying it, so this information _needs_ to be 100% accurate.

What will be beyond confusing to users (and me) is the following situation:
- There's a 512x512 vanilla game texture. Texplorer thinks the mips only goes down to 4x4, which means it (apparently) has 8 mips and not the 10 it should have.
- Does it then display this as 10 or 8? 
- Does this number include these "null" mips or not? If so, then, is it 6 or 4 mips?
- They then extract the texture and edit in GIMP, which includes upscaling it to 1024. They check "Generate mipmaps" when they save the texture. Their new texture now has 11 mips, b/c the auto-generating mips feature of these programs will generate them down to 1x1..
- They re-open the DDS in GIMP to check for integrity and they notice there are 11 mips and not 9 or 7 or 5 (or 10 or 8 or 6 or 4).
- They are now confused as to how many mips the replacement texture should have.

If I'm misunderstanding what K is saying and this can't happen, then he should let me know. But, this is exactly what I see happening -- and frequently -- from his above explanation.

If we cannot be certain the data we are providing about which mips are present _is accurate_, then we should not present the information at all. Otherwise, it's misleading. We already have a system in place where when mips are present for a game texture, then the mips for the replacement texture must go down to 1x1. This is literally all anyone needs to know.

Don't get me wrong, @CreeperLava, I am all about thoroughness and having a complete source of information. I think you know that. But, the information has to be correct, consistent across tools, adhere to conventions used in other programs, etc.

**Comment by starkgate** [comment]
Hmm, I suppose you're right, there's no real solution to this problem. We'll see how it goes in the future, but I think I will have to go back to older revisions from time to time for features like MIPs extraction and the likes, that have been removed.

**Comment by KFreon** [comment]
Texplorer is 100% accurate for the game (as far as we know), it's just that the game doesn't necessarily play by the same rules. 
Texplorer is also supposed to be the beez-neez of texture information for the toolset. This thing is supposed to provide as much information on a texture as possible without digging into the code yourself. 
Your example of a user checking to see if something worked is reasonably confusing, BUT no user would do this, only modders. AND like you said, they wouldn't know what mips are to be confused as to why they aren't the same number, AND it isn't there atm anyway. 
The mip count is for us, not users, and users shouldn't be fiddling around in Texplorer.
The only reason Texplorer is still accessible by users is because I got convinced years ago to keep it because people might want to find textures.
I should have removed all functionality from it ages ago and made it a viewer only.

**Comment by ghost** [comment]
@KFreon -- I must be missing something with your quote here, b/c it's not making sense:

> Texplorer is 100% accurate for the game (as far as we know), it's just that the game doesn't necessarily play by the same rules. 

Texplorer is just conveying the information as stored in the game files, so... how can they not play by the same rules? Texplorer's rules by default should be _the game's rules_.

Please keep in a mind a few things with this whole discussion....
1. With the advent of DLC mods, new custom texture TFCs and PCCs, the editing functionalities of Texplorer are now more necessary than ever; they very much need to stick around. I know you dislike a lot of things about the tool, but it's editing features are what's made Enhanced Galaxy Mod, BackOff, AVPen's armor mods, and a lot of other mods possible. TPF Tools is nice for just installing textures, but like you said, Texplorer is a much more powerful tool for modders. I, for one, don't want to have to switch between the two tools when it comes to implementing a new texture into the game (not replacing, _creating_). Right now, the workflow for that entire process is inside Texplorer, which is ideal.
2. As a modder who understands this stuff pretty well, _I_ would be confused by the system of mip display numbers you described (null/not null, whether they go down to 1x1, etc). I get that you think that all modders know this stuff, but that's not always the case. We all are learners at some point, even as modders. Texture modding is also a very specialized area, and with DLC mods, more and more mod authors -- who know nothing about textures -- are starting to dabble in the field. Let's make it easier for them :)
3. There are A LOT of outdated TPFs out there. Yes, TPF Tools can autofix some (even most) of these, but it's not foolproof. Every texture user will occasionally need to dive into Texplorer to check on format and mips (well, previously), and extract the vanilla texture for comparison. I did _a ton_ of this (wearing my "user hat") when attempting to perma-mod ME1 and ME2 with textures, and I doubt I'm the only one.

Texplorer is a great tool. A fantastic one. Editing functionalities and all. It's also really not difficult to use; not for modders or users :)

All this talk just about how to implement a mips field, eek, lol.

**Comment by KFreon** [comment]
That's what I mean: Texplorer plays by the games rules, but not by anyone elses rules. i.e. GIMP, ImageEngine and every other thing that handles DDS' will have different mipcounts to the game, since it's done tricky things to help itself out.
1. Fair enough.
2. It would say: **Mips: 12**. I guess what I was getting at there was that if you know all about mips, you now know more about the texture, but if you don't know about mips, it doesn't really matter, I mean why does it matter for a modder to know precisely how many there are?
3. Hmm fair enough I guess...

Thinking about it, how about the mip field look like this:
- **Mips:** Yes (12)

That way gift gets her "yes/no" and we get our number.

**Comment by starkgate** [comment]
> Mips: Yes (12)
> That way gift gets her "yes/no" and we get our number.

A big YES to this suggestion !

**Comment by ghost** [comment]
@KFreon -- If it's for a 2048x2048 texture, then yeah, I'd say that's great. 12 is what it should be, and that's exactly what TPF Tools will say, as well -- assuming the new modded texture is formatted correctly.

> That way gift gets her "yes/no" and we get our number.

Lol, always working to please me, K. Such a gentleman.

But no, if 12 = 2048p, that's perfect.

**Comment by KFreon** [comment]
What I meant was: 
- **Mips:** (number of detected mips) so for a 2048 texture, that could only be 5 if the game texture is weird.

**Comment by ghost** [comment]
> What I meant was:
> 
> Mips: (number of detected mips) so for a 2048 texture, that could only be 5 if the game texture is weird.

That's where you lose me, though. I guess as long as it says "yes", it's okay, but this is going to be so hard to explain on the wiki. 

We've already been very explicit that 2048=12 mips. There's a pretty table and everything that provides a list of all mip sizes and numbers for a specific "parent" texture size. We've drilled "proper mips" into folks' heads. By placing a 5 next to a 2048 texture, I really think it's going to profoundly confuse people.  They'll think "Well, the vanilla game texture has 5 mips, then my new texture should also have 5 mips".

**Comment by KFreon** [comment]
I can change it to something like:
- **Mips:** Yes (5 detected/valid) 
  Something like that?

I'm still dubious as to whether it matters, but you're closer to it than I am, I'l defer to you.

**Comment by ghost** [comment]
@KFreon -- Uuuugh. 

Okay, let me ask you this: If the toolset only needs 5 mips for this hypothetical texture, why do we have to provide more when replacing it?

That sort of gets to the crux of the issue.

I honestly don't think that whether they are "valid" or not matters. If the number of mips we have to use to run a successful replacement isn't going to change based upon detected/valid, then that information is not only confusing, but irrelevant.

Let me wait until you respond to my question above before I say more and make up my mind, though :)

**Comment by KFreon** [comment]
@giftfish - Technically you don't, but the toolset, along with every other tool ever, wants properly formatted DDS'. 
The game doesn't seem to follow the rules set out by Microsoft's DDS spec, so it doesn't need the correct number of mipmaps for its size.
So the game has a 1024x1024 texture with only 3 mips, and when you go to replace it, you can only provide a DDS with 1 mip (i.e. full size), or one with all 11 mips. No tool I know of does a half job of this.
- So Texplorer reports on what the game texture is saying: "I'm 1024, but I've only got 5 mips".
- You want to add/replace with a different texture and the engine goes: "you're 2048x2048, so you should have 12 mips. Yep header says 12 but let me check the length. Ok length is good, you are indeed a properly formatted DDS." and off it goes replacing things.

If you could  get a tool to only give you 5 mips for a 1024 image, the checks could be removed and that would work.

tl;dr. Texplorer imo should report on what the game says, but it needs a proper DDS to do the replacing.
Hopefully that answers your question.

**Comment by ghost** [comment]
@KFreon -- Okay, that is ridiculously convoluted, but your explanation makes sense, and I think I can parse it on the wiki.

The rule will be this:

> If a texture has mips (Mipmaps=Yes), then replacement textures must also have mips. 
> 
> Additional Explanation: Replacement textures must _always_ have the full compliment of mipmaps, as automatically generated by any image program. This means new textures will often have more mips than the original, even at the same resolution. 

Then I'll direct folks to the handy table that shows how many mips at each resolution. 

How does that sound? If good, then, yes, your format of...
- Mipmaps: Yes (#)

...will work just fine.

Am I correct in assuming that a GUI texture would look like this?
- Mipmaps: No

**Comment by KFreon** [comment]
That sounds good, and a UI texture or one that doesn't have mipmaps will look thusly:
- **Mips:** No (1).

**Comment by ghost** [comment]
Ah. Got it. That makes sense, and I'll be sure it's explicit in the Texplorer article.

**Comment by ghost** [comment]
@KFreon -- This field looks great. I can't verify the numbers of course (I don't think there's any way to do this, short of comparing to revs with the old extract dialog), but all the y/n I see make sense.

Nice job.

---

Verified in r747.
